File: dbflow-tests/src/test/java/com/raizlabs/android/dbflow/test/structure/caching/MultipleCacheableModel.java
package com.raizlabs.android.dbflow.test.structure.caching;

import android.support.annotation.NonNull;

import com.raizlabs.android.dbflow.annotation.ForeignKey;
import com.raizlabs.android.dbflow.annotation.ForeignKeyReference;
import com.raizlabs.android.dbflow.annotation.MultiCacheField;
import com.raizlabs.android.dbflow.annotation.PrimaryKey;
import com.raizlabs.android.dbflow.annotation.Table;
import com.raizlabs.android.dbflow.structure.BaseModel;
import com.raizlabs.android.dbflow.structure.cache.IMultiKeyCacheConverter;
import com.raizlabs.android.dbflow.test.TestDatabase;
import com.raizlabs.android.dbflow.test.structure.TestModel1;

/**
 * Description:
 */
@Table(database = TestDatabase.class, cachingEnabled = true)
public class MultipleCacheableModel extends BaseModel {

    @MultiCacheField
    public static final IMultiKeyCacheConverter<String> multiKeyCacheModel = new IMultiKeyCacheConverter<String>() {

        @Override
        @NonNull
        public String getCachingKey(@NonNull Object[] values) {
            return "(" + values[0] + "," + values[1] + ")";
        }
    };

    @PrimaryKey
    double latitude;

    @PrimaryKey
    double longitude;

    @ForeignKey(references = {@ForeignKeyReference(columnName = "associatedModel",
            columnType = String.class, foreignKeyColumnName = "name", referencedFieldIsPackagePrivate = true)})
    TestModel1 associatedModel;

}

File: dbflow-tests/src/test/java/com/raizlabs/android/dbflow/test/structure/caching/MultipleCacheableModel.kt
package com.raizlabs.android.dbflow.test.structure.caching

import com.raizlabs.android.dbflow.annotation.*
import com.raizlabs.android.dbflow.structure.BaseModel
import com.raizlabs.android.dbflow.structure.cache.IMultiKeyCacheConverter
import com.raizlabs.android.dbflow.test.TestDatabase
import com.raizlabs.android.dbflow.test.structure.TestModel1

/**
 * Description:
 */
@Table(database = TestDatabase::class, cachingEnabled = true)
class MultipleCacheableModel : BaseModel() {

    @PrimaryKey
    var latitude: Double = 0.toDouble()

    @PrimaryKey
    var longitude: Double = 0.toDouble()

    @ForeignKey(references = arrayOf(ForeignKeyReference(columnName = "associatedModel",
        columnType = String::class, foreignKeyColumnName = "name",
        referencedFieldIsPrivate = true)))
    var associatedModel: TestModel1? = null

    companion object {

        @JvmField
        @MultiCacheField
        val multiKeyCacheModel: IMultiKeyCacheConverter<String> = IMultiKeyCacheConverter { values -> "(" + values[0] + "," + values[1] + ")" }
    }

}
